code/asteroid/asteroid.cpp

```diff
@@ -733,52 +780,75 @@ static void asteroid_aim_at_target(object *objp, object *asteroid_objp, float de
 }
 
 /**
- * Call once per frame to maybe throw an asteroid at a ship.
+* Sanitizes the Asteroid targets list removing ships that have died/become invalid etc
+*/
+static void sanitize_asteroid_targets_list() {
+	for (size_t i = 0; i < Asteroid_targets.size();) {
+		object* target_objp = &Objects[Asteroid_targets[i].objnum];
+		if (target_objp->type != OBJ_SHIP || target_objp->signature != Asteroid_targets[i].signature) {
+			Asteroid_targets[i] = Asteroid_targets.back();
+			Asteroid_targets.pop_back();
+		}
+		else  // if we needed to cull we should not advance because we just moved a new asteroid target into this spot
+			i++;
+	}
+}
+
+/**
+ * Call once per frame to maybe throw some asteroids at one or more ships.
  *
- * @param count asteroids already targeted on
  */
-static void maybe_throw_asteroid(int count)
+static void maybe_throw_asteroid()
 {
-	if (!timestamp_elapsed(Next_asteroid_throw)) {
-		return;
-	}
 
-	if (Asteroid_throw_objnum == -1) {
-		return;
-	}
+	for (asteroid_target& target : Asteroid_targets) {
+		if (!timestamp_elapsed(target.throw_stamp))
+			continue;
 
-	nprintf(("AI", "Incoming asteroids: %i\n", count));
+		object* target_objp = &Objects[target.objnum];
+		if (!asteroid_is_ship_inside_field(&Asteroid_field, &target_objp->pos, target_objp->radius))
+			continue;
+		
+		// This should've been greater equal, but alas...
+		if (target.incoming_asteroids > The_mission.ai_profile->max_incoming_asteroids[Game_skill_level])
+			continue;
 
-	if (count > The_mission.ai_profile->max_incoming_asteroids[Game_skill_level])
-		return;
+		nprintf(("AI", "Incoming asteroids to %s: %i\n", Ships[target_objp->instance].ship_name, target.incoming_asteroids));
 
-	Next_asteroid_throw = timestamp(1000 + 1200 * count/(Game_skill_level+1));
+		target.throw_stamp = timestamp(1000 + 1200 * target.incoming_asteroids /(Game_skill_level+1));
 
-	ship_obj	*so;
-	for ( so = GET_FIRST(&Ship_obj_list); so != END_OF_LIST(&Ship_obj_list); so = GET_NEXT(so) ) {
-		object *A = &Objects[so->objnum];
-		if (so->objnum == Asteroid_throw_objnum) {
-			int subtype = Random::next(NUM_DEBRIS_POFS);
-			while (Asteroid_field.field_debris_type[subtype] == -1) {
-				subtype = (subtype + 1) % NUM_DEBRIS_POFS;
+		int counter = Random::next(Asteroid_field.num_used_field_debris_types);
+		int subtype = -1;
+		for (int i = 0; i < NUM_DEBRIS_POFS; i++) {
+			if (Asteroid_field.field_debris_type[i] >= 0) {
+				if (counter == 0) {
+					subtype = i;
+					break;
+				}
+				else
+					counter--;
```

Comment:
I was thinking that this selection logic could be put in its own function, but since it's only used in two places (remember the "rule of three") this is probably ok.